### PR TITLE
Add validators to request DTOs

### PR DESCRIPTION
**Pull request by connernilsen**
Are there any other validations that should be put in, or any that should be removed? I got most of them, but I wasn't sure about some. 

Also is this something you want, or would you just prefer to handle the cases that come up?

### CODE REVIEW

File: api/src/main/java/com/codeforcommunity/dto/auth/NewUserRequest.java [hunk 1]
```diff
@@ -29,14 +32,27 @@ public NewUserRequest(
...
+    if (password == null) {
+      throw new MalformedParameterException("Password");
+    }
+    if (firstName == null) {
+      throw new MalformedParameterException("First name");
+    }
+    if (lastName == null) {
+      throw new MalformedParameterException("Last name");
+    }
+    if (phoneNumber == null) {
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/dto/auth/NewUserRequest.java:49]: Phone number is not requiired

File: api/src/main/java/com/codeforcommunity/dto/userEvents/requests/CreateEventRequest.java [hunk 2]
```diff
@@ -37,4 +39,19 @@ public String getThumbnail() {
...
+
+  @Override
+  public void validate() {
+    if (title == null) {
+      throw new MalformedParameterException("Title");
+    }
+    if (spotsAvailable == null) {
+      throw new MalformedParameterException("Spots available");
+    }
+    // TODO: can the thumbnail be null?
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/dto/userEvents/requests/CreateEventRequest.java:51]: yes

File: api/src/main/java/com/codeforcommunity/dto/announcements/GetAnnouncementsRequest.java [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
+import com.codeforcommunity.api.ApiDto;
 import com.codeforcommunity.exceptions.MalformedParameterException;
 import java.sql.Timestamp;
 import java.util.Date;
 
-public class GetAnnouncementsRequest {
+public class GetAnnouncementsRequest implements ApiDto {
 
   private Timestamp startDate; // optional
   private Timestamp endDate; // optional
```
Comment (connernilsen) [inline api/src/main/java/com/codeforcommunity/dto/announcements/GetAnnouncementsRequest.java:11]: Fields in this class are noted as optional, but can they be nullable?

File: api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java [hunk 4]
```diff
@@ -1,6 +1,9 @@
 package com.codeforcommunity.dto.announcements;
 
-public class PostAnnouncementRequest {
+import com.codeforcommunity.api.ApiDto;
+import com.codeforcommunity.exceptions.MalformedParameterException;
+
+public class PostAnnouncementRequest implements ApiDto {
 
   private String title;
   private String description;
```
Comment (connernilsen) [inline api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java:9]: I'm checking if title is empty, should I also ensure description can't be empty?

File: api/src/main/java/com/codeforcommunity/dto/userEvents/components/EventDetails.java [hunk 5]
```diff
@@ -1,8 +1,9 @@
 package com.codeforcommunity.dto.userEvents.components;
 
+import com.codeforcommunity.api.ApiDto;
 import java.sql.Timestamp;
 
-public class EventDetails {
+public class EventDetails implements ApiDto {
   private String description;
   private String location;
```
Comment (connernilsen) [inline api/src/main/java/com/codeforcommunity/dto/userEvents/components/EventDetails.java:8]: Should `description` and `location` be able to be empty?

File: api/src/main/java/com/codeforcommunity/dto/announcements/GetAnnouncementsRequest.java (hunk 3 above)
Comment (connernilsen) [reply api/src/main/java/com/codeforcommunity/dto/announcements/GetAnnouncementsRequest.java:8]: this one isn't being used :)

File: api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java (hunk 4 above)
Comment (connernilsen) [reply api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java:10]: The description thing applies to a lot of other things too...

File: api/src/main/java/com/codeforcommunity/dto/checkout/LineItemRequest.java [hunk 6]
```diff
@@ -19,4 +23,17 @@ public Integer getEventId() {
...
+  @Override
+  public List<String> validateFields(String fieldPrefix) {
+    String fieldName = fieldPrefix + "line_item_request.";
+    List<String> fields = new ArrayList<>();
+    if (eventId == null || eventId < 1) {
+      fields.add(fieldName + "event_id");
+    }
+    if (quantity == null || quantity < 1) {
+      fields.add(fieldName + "quantity");
+    }
```
Comment (connernilsen) [inline api/src/main/java/com/codeforcommunity/dto/checkout/LineItemRequest.java:36]: Can this be 0?

File: api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java (hunk 4 above)
Comment (liammoyn) [reply api/src/main/java/com/codeforcommunity/dto/announcements/PostAnnouncementRequest.java:10]: I think allowing descriptions to be empty is fine

File: api/src/main/java/com/codeforcommunity/dto/userEvents/components/EventDetails.java (hunk 5 above)
Comment (liammoyn) [reply api/src/main/java/com/codeforcommunity/dto/userEvents/components/EventDetails.java:12]: location no, description yes

File: api/src/main/java/com/codeforcommunity/dto/checkout/LineItemRequest.java (hunk 6 above)
Comment (liammoyn) [reply api/src/main/java/com/codeforcommunity/dto/checkout/LineItemRequest.java:36]: No

File: api/src/main/java/com/codeforcommunity/api/ApiDto.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    throw new MalformedParameterException(builder.toString());
+  }
+
+  /**
+   * Checks to see if email is valid.
+   *
+   * @param email the email to check.
+   * @return a boolean representing whether this email is not valid.
+   */
+  public boolean emailInvalid(String email) {
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/api/ApiDto.java:82]: Can these methods be protected?

Comment (connernilsen) [reply api/src/main/java/com/codeforcommunity/api/ApiDto.java:82]: Sure

**Pull request merged**